### PR TITLE
OGC 3DTiles Probe for #381

### PR DESCRIPTION
**Pull request by jochemthart1**
This PR adds a probe for OGC 3D Tile services. (#381 )

It requests the tileset.json, retrieves the tile data url from the json and requests the tile data url as well.

Currently, it is written as one probe that checks both the tileset.json and the data url. Another probe could be added if only the tileset.json should be checked.

Both requests are checked for http4/500 errors using the GHC HttpStatusNoError check.

I had to make small changes to `models.py` and `healthcheck.py` to make it work differently to other "OGC:" service probes. (similar to `OGC:WFS3`)

### CODE REVIEW

File: GeoHealthCheck/plugins/probe/ogc3d.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            self.response = Probe.perform_get_request(self, tile_url)
+            self.run_checks()
+        except requests.exceptions.RequestException as e:
+            msg = "Request Err: Error requesting tileset.json %s %s" \
+                % (e.__class__.__name__, str(e))
+            self.result.set(False, msg)
+
+        # Get data url from tileset.json and request tile data
+        try:
+            data_uri = self.get_3d_tileset_content_uri(self.response.json())
```
Comment (justb4) [inline GeoHealthCheck/plugins/probe/ogc3d.py:44]: Optional: you may want to check for `None` return value from `get_3d_tileset_content_uri()` (or have that funtion throw exception), but ok, will be caught in Exception. For analysing errors always good to be specific on the error i.s.o. 'has value None` error.

File: GeoHealthCheck/plugins/probe/ogc3d.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            data_url = url_base + '/' + data_uri
+            self.log('Requesting: %s url=%s' % (self.REQUEST_METHOD, data_url))
+            self.response = Probe.perform_get_request(self, data_url)
+            self.run_checks()
+        except requests.exceptions.RequestException as e:
+            msg = "Request Err: Error requesting tile data %s %s" \
+                % (e.__class__.__name__, str(e))
+            self.result.set(False, msg)
+
+    def check_response(self):
```
Comment (justb4) [inline GeoHealthCheck/plugins/probe/ogc3d.py:54]: Think this function is not required anymore (?), as `'GeoHealthCheck.plugins.check.checks.HttpStatusNoError'` is the default check.

**Review by justb4 — CHANGES_REQUESTED** [review]
Good to see this! You may want to add more specific Probe/Check stuff later. Only minor comments:

* make more specific that this Probe is for OGC 3D Tiles. There are and will be more OGC 3D Services like OGC 3D Portrayal (ok hardly used). Specific in identifier naming:

`OGC:3D` --> `OGC:3DTiles`
file `ogc3d.py` --> `ogc3dtiles.py`

**Comment by justb4** [comment]
> I had to make small changes to models.py and healthcheck.py to make it work differently to other "OGC:" service probes. (similar to OGC:WFS3)
Yes, that is always the case for new service types!

**Comment by fsteggink** [comment]
This PR is meant for issue #381

File: GeoHealthCheck/plugins/probe/ogc3d.py (hunk 2 above)
Comment (jochemthart1) [reply GeoHealthCheck/plugins/probe/ogc3d.py]: Removed the redundant `check_response()` function and changed the names from ogc3d to ogc3dtiles and from OGC:3D to OGC:3DTiles

File: GeoHealthCheck/plugins/probe/ogc3dtiles.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        try:
+            tile_url = url_base + '/tileset.json'
+            self.log('Requesting: %s url=%s' % (self.REQUEST_METHOD, tile_url))
+            self.response = Probe.perform_get_request(self, tile_url)
+            self.run_checks()
+        except requests.exceptions.RequestException as e:
+            msg = "Request Err: Error requesting tileset.json %s %s" \
+                % (e.__class__.__name__, str(e))
+            self.result.set(False, msg)
+
```
Comment (justb4) [inline GeoHealthCheck/plugins/probe/ogc3dtiles.py:41]: Better add `return` here as `/tileset.json` already gives `Exception`, no use going on.

**Review by justb4 — APPROVED** [review]
Ok, only one minor in Probe file (see comment there). You can leave it or add the `return`.

For testing, maybe later: Looking for public 3DTiles services, found:
https://3dbag.nl/download/3dtiles/v210908_fd2cee53/lod22/tileset.json though this URL does not seem to be stable.

**Comment by justb4** [comment]
@jochemthart1 let me know if you are ready for merge, my comments are minor.

File: GeoHealthCheck/plugins/probe/ogc3dtiles.py (hunk 3 above)
Comment (jochemthart1) [reply GeoHealthCheck/plugins/probe/ogc3dtiles.py:43]: You're right. Added return here.

Comment (jochemthart1) [reply GeoHealthCheck/plugins/probe/ogc3dtiles.py:45]: This was actually a mistake where the probe would only look directly under the root for a data url. I changed it to use recursion to "dig deeper" into the tileset.json to find the tile data url. The url you found https://3dbag.nl/download/3dtiles/v210908_fd2cee53/lod22/tileset.json does now give back a "True" result.

**Comment by jochemthart1** [comment]
@justb4 This is now ready for merge. I did implement some changes based on your review.

File: GeoHealthCheck/plugins/probe/ogc3dtiles.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import requests
+
+
+class OGC3DTiles(Probe):
+    """
+    OGC3DTiles
+    """
+
+    NAME = 'GET Tileset.json and tile data'
+    DESCRIPTION = 'OGC3DTiles'
```
Comment (justb4) [inline GeoHealthCheck/plugins/probe/ogc3dtiles.py:11]: PLease provide more extended, one line description. This text will appear in the Resource Edit UI, next to "Probe" in italics.

File: GeoHealthCheck/plugins/probe/ogc3dtiles.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            url_base = url_base.split('/tileset.json')[0]
+
+        # Request tileset.json
+        try:
+            tile_url = url_base + '/tileset.json'
+            self.log('Requesting: %s url=%s' % (self.REQUEST_METHOD, tile_url))
+            self.response = Probe.perform_get_request(self, tile_url)
+            self.log('test1')
+            self.run_checks()
+            self.log('test2')
```
Comment (justb4) [inline GeoHealthCheck/plugins/probe/ogc3dtiles.py:38]: Pls remove excessive logging/debug statements.

File: GeoHealthCheck/plugins/probe/ogc3dtiles.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                % (e.__class__.__name__, str(e))
+            self.result.set(False, msg)
+            # If error occurs during request of tileset.json, no use going on
+            return
+
+        # Get data url from tileset.json and request tile data
+        try:
+            tile_root = self.response.json()['root']
+            data_uri = self.get_3d_tileset_content_uri(tile_root)
+            print('DATA URI', data_uri)
```
Comment (justb4) [inline GeoHealthCheck/plugins/probe/ogc3dtiles.py:50]: Pls remove `print()` line.

**Comment by justb4** [comment]
@jochemthart1 some more patience: see my minor comments above.

**Review by justb4 — APPROVED** [review]

**Comment by justb4** [comment]
@jochemthart1 almost there, still a flake8 error:
```
./GeoHealthCheck/plugins/probe/ogc3dtiles.py:11:80: E501 line too long (84 > 79 characters)
Error: Process completed with exit code 1.
````

**Review by justb4 — APPROVED** [review]

**Comment by justb4** [comment]
@jochemthart1 ok, here we go, thanks for patience!

**Comment by justb4** [comment]
And on demo server: https://demo.geohealthcheck.org/resource/412?lang=en

**Pull request merged**